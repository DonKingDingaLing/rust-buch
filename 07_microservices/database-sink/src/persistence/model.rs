use diesel::{self, Identifiable, Queryable, Insertable, AsChangeset};
use serde::*;
use crate::persistence::schema::requests;
use std::convert::TryFrom;

#[table_name = "requests"]
#[derive(
Serialize,      //
Deserialize,    //
Clone,          //
Debug,          //
PartialEq,      //
Identifiable,   //
Queryable,      //
Insertable,     //
AsChangeset,    //
)]
pub struct ContactRequest {
    pub id: i32,
    pub email: String,
    pub message: String,
}

#[derive(Insertable, Serialize, Deserialize)]
#[table_name="requests"]
pub struct NewContactRequest {
    pub email: String,
    pub message: String,
}

impl TryFrom<&[u8]> for NewContactRequest{
    type Error = ();
    fn try_from(value: &[u8]) -> Result<Self, Self::Error> {
        serde_json::from_slice(value).map_err(|_e|())
    }
}