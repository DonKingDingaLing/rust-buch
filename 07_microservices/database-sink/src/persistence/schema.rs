table! {
    requests (id) {
        id -> Int4,
        email -> Varchar,
        message -> Varchar,
    }
}
