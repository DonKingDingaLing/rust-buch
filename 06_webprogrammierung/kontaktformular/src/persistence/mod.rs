pub mod request_repo;
use rocket_sync_db_pools::diesel::SqliteConnection;
use rocket_sync_db_pools::database;


#[database("sqlite_db")]
pub struct DbCon(SqliteConnection);