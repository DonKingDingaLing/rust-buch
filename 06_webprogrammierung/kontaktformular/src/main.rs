#[macro_use]
extern crate rocket;

#[macro_use]
extern crate diesel;

use persistence::DbCon;
use crate::auth::SessionStorage;
use rocket_dyn_templates::Template;

mod model;
mod persistence;
mod routes;
mod schema;
mod auth;
mod catchers;

/**
    Start the rocket
    The `launch` macro makes this method asynchronous and spawns it in a worker.
    Further, the macro takes care, that the function is called from the main function.

    The SessionStore is used for storing Sessions for cookie-based authentication and acts as a
        demonstration of how to use managed state. Since Rocket requires its managed State to be
        thread-safe, session-store uses an RwLock internally an can be used from any thread.
    The PgDbCon fairing acts as a wrapper around diesels PgConnection but has a annotation from
        rocket, so that it can be configured by the Rocket.toml configuration file.
    The routes mounted are implemented in their respective module files for maintainability.
    The single implemented custom error catcher shal demonstrate how this feature works in rocket.
*/

#[launch]
fn launch() -> _ {

    let session_store = SessionStorage::new();

    rocket::build()
        .attach(DbCon::fairing())
        .attach(Template::fairing())
        .manage(session_store)
        .mount(
            "/public/",
            routes![
                routes::public::index_json,
                routes::public::index_template,
                routes::public::index_form,
            ],
        )
        .mount(
            "/admin/",
            routes![
                routes::admin::list_requests,
                routes::admin::get_request_by_id,
                routes::admin::delete_request_by_id,
                routes::admin::update_request,
                routes::admin::create_request,
                routes::admin::list_requests_template,
            ],
        )
        .mount(
            "/auth/",
            routes![
                auth::login_template,
                auth::logout_template,
                auth::login_form,
            ],
        )

        /*
         * Register is used to register catchers in the Rocket instance.
         * Catchers are handlers that are invoked, when errors occur.
         * Catchers will be invoked, if no other handlers match (e.g. 404) or some handler
         * returned some Ourtcome other than Success.
         * Returning an Outcome::Failure or Outcome::Forward and handling them with catchers
         * is extremely powerfull, since one can externalize Error handling from the "real"
         * handlers.
         *
         * The catchers! macro creates a Vec of Catchers with their respective routes.
         * The routes for each catcher can be defined by using the mcaro #[catch(...)]
         */
        .register(
            "/",catchers![catchers::unauthorized]
        )
}

#[cfg(test)]
mod test {
    use super::launch;
    use rocket::local::blocking::Client;
    use rocket::http::Status;

    #[test]
    fn hello_world() {
        let client = Client::tracked(launch()).expect("valid rocket instance");
        let mut response = client.get("/public").dispatch();
        assert_eq!(response.status(), Status::Ok);
        insta::assert_yaml_snapshot!(response.into_string().unwrap());
    }
}
