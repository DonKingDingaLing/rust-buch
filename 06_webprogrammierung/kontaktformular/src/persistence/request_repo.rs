use crate::diesel::QueryDsl;
use crate::diesel::RunQueryDsl;
use crate::model::{ContactRequest, NewContactRequest};
use crate::persistence::DbCon;
use crate::schema::requests;
use diesel::connection::Connection;
use diesel::result::Error;
use diesel::QueryResult;
use diesel::{self, prelude::*};
use crate::schema::requests::dsl::{requests as all_requests};


/**
    Returns all ContactRequests in the database.
    If there is a Database Error of any kind, we map it to its string representation
 */
pub async fn get_all(con: &DbCon) -> Result<Vec<ContactRequest>, Error> {
    let result = con.run(|c| {
        requests::table
            .load::<ContactRequest>(c)
    }).await;
    return result;
}

/**

*/
pub async fn insert(con: &DbCon, new_request: NewContactRequest) -> Result<ContactRequest, String> {
    use diesel::prelude::*;
    use diesel::select;

    let result = con.run(move |c| {
        /*
            We explicitly start a transaction here, since SQLite does not support returning the inserted
            elements. On a real database, you can use `get_result` (commented out below) to get a Result of type
            `Result<ContactRequest, Error>` here.
         */
        let query_result = c.transaction::<_, Error, _>(|| {
            diesel::insert_into(requests::table)
                .values(&new_request)
                //.get_result(&con.0);
                .execute(c)?;

            requests::table
                .order(requests::id.desc())
                .limit(1)
                .load(c)?
                .into_iter()
                .rev()
                .next()
                .ok_or(Error::NotFound)
        });
        query_result.map_err(|e| e.to_string())
    }).await;
    return result;
}

/* // Would we only pg, the obove function could be condensed to this
pub async fn insert_pg_only(con: &DbCon, new_request: NewContactRequest) -> Result<ContactRequest, String> {
    use diesel::prelude::*;
    use diesel::select;
    con.run(move |c| {
        diesel::insert_into(requests::table)
            .values(&new_request)
            .get_result(c)
            .map_err(|e| e.to_string())
    }).await
}
*/

/**

*/
pub async fn get_by_id(con: &DbCon, request_id: i32) -> Result<Option<ContactRequest>,String> {
    con.run(move |c| {
        requests::table
            .find(request_id)
            .first(c)
            .map(|x| Some(x))         // TODO Das ergibt so keinen sinn
            .map_err(|e| e.to_string())
    }).await
}

/**

*/
pub async fn delete_by_id(con: &DbCon, request_id: i32) -> Result<usize, String> {
    let result = con.run(move |c| {
        diesel::delete(requests::table.find(request_id))
            .execute(c)
            .map(|num| num)
            .map_err(|e| e.to_string())
    }).await;
    return result;

}

/**

*/
pub async fn update(con: &DbCon, id: i32, updated_request: ContactRequest) -> Result<ContactRequest, String> {
    if updated_request.id != id {
        return Err("The id of a request cannot be changed".to_string())
    }

    /*
    If we used only postgres, the code of this method could be condensed to the following
    diesel::update(requests::table.find(updated_request.id))
        .set(updated_request)
        .get_result(&con.0)
        .map_err(|e| e.to_string())
     */

    use diesel::prelude::*;
    use diesel::select;

    con.run(|c| {
        /*
            We explicitly start a transaction here, since SQlite does not support returning the inserted
            elements. On a real database, you can use `get_result` (commented out below) to get a Result of type
            `Result<ContactRequest, Error>` here.
         */
        let query_result = c.transaction::<_, Error, _>(|| {
            diesel::update(requests::table.find(updated_request.id))
                .set(updated_request)
                //.get_result(&con.0);
                .execute(c)?;

            requests::table
                .order(requests::id.desc())
                .limit(1)
                .load(c)?
                .into_iter()
                .rev()
                .next()
                .ok_or(Error::NotFound)
        });
        query_result.map_err(|e| e.to_string())
    }).await
}
