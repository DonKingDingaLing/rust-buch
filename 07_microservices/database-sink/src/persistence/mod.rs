use diesel::{PgConnection, Connection};
use crate::MyConfig;
use crate::persistence::model::{ContactRequest, NewContactRequest};
use log::{trace, debug};

use schema::requests;
use diesel::prelude::RunQueryDsl;

pub mod model;
pub mod schema;


pub (crate) fn connect_db(conf: &MyConfig) -> PgConnection {
    trace!("About to connect to DB");

    let con = match PgConnection::establish(&conf.db_url) {
        Ok(con) => Some(con),
        Err(e) =>  {   debug!("Error: {}", e);  None }
    };

    trace!("Finished to connect to DB");
    return con.unwrap();
}

pub (crate) fn save(con: &PgConnection, request: NewContactRequest)
    -> Result<ContactRequest,String> {
    diesel::insert_into(requests::table)
        .values(&request)
        .get_result(con)
        .map_err(|e| format!("DB Failure: {}",e))
}