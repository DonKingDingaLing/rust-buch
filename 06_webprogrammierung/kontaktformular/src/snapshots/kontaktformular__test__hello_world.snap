---
source: src/main.rs
expression: response.into_string().unwrap()

---
"<!DOCTYPE html>\n<html>\n  <body>\n      <form  action=\"/public/\" method=\"post\">\n          <label for=\"email\">E-Mail:</label><br>\n          <input type=\"email\" id=\"email\" name=\"email\"><br>\n          <label for=\"message\">Nachricht:</label><br>\n          <input type=\"text\" id=\"message\" name=\"message\">\n          <input type=\"submit\" value=\"Submit\">\n      </form>\n  </body>\n</html>"

